libraries/chain/chain_controller.cpp

```diff
@@ -872,51 +878,84 @@ void chain_controller::validate_expiration(const Transaction& trx) const
 } FC_CAPTURE_AND_RETHROW((trx)) }
 
 uint32_t chain_controller::_transaction_message_rate(uint32_t now, uint32_t last_update_sec, uint32_t rate_limit_time_frame_sec,
-                                                     uint32_t rate_limit, uint32_t previous_rate, const char* type, const AccountName& name)
+                                                     uint32_t rate_limit, uint32_t previous_rate, rate_limit_type type, const AccountName& name)
 {
    const auto delta_time = now - last_update_sec;
    uint32_t message_count = 1;
    if (delta_time <= rate_limit_time_frame_sec)
    {
       message_count += ( ( ( rate_limit_time_frame_sec - delta_time ) * fc::uint128( previous_rate ) )
                      / rate_limit_time_frame_sec ).to_uint64();
-      EOS_ASSERT(message_count <= rate_limit, tx_msgs_exceeded,
-                 "Rate limiting ${type}=${name} messages sent, ${count} exceeds ${max} messages limit per ${sec} seconds. Wait 1 second and try again",
-                 ("type",type)
-                 ("name",name)
-                 ("count",message_count)
-                 ("max",rate_limit)
-                 ("sec", rate_limit_time_frame_sec));
+#define RATE_LIMIT_ASSERT(tx_msgs_exceeded, type_str) \
+      EOS_ASSERT(message_count <= rate_limit, tx_msgs_exceeded, \
+                 "Rate limiting ${type} account=${name} messages sent, ${count} exceeds ${max} messages limit per ${sec} seconds. Wait 1 second and try again", \
+                 ("type",type_str) \
+                 ("name",name) \
+                 ("count",message_count) \
+                 ("max",rate_limit) \
+                 ("sec", rate_limit_time_frame_sec))
+      switch (type)
+      {
+      case authorization_account:
+         RATE_LIMIT_ASSERT(tx_msgs_auth_exceeded, "authorization");
+         break;
+      case code_account:
+         RATE_LIMIT_ASSERT(tx_msgs_code_exceeded, "code");
+         break;
+      }
```

Comment:
Add a default case for safety, "unknown"

Reply:
What type of exception should it throw?

Reply:
How about just FC_ASSERT() instead.